apps/nouns-camp/src/components/action-dialog.js

```diff
@@ -801,6 +823,127 @@ const formConfigByActionType = {
     }),
     Component: StreamingPaymentActionForm,
   },
+  "treasury-noun-transfer": {
+    title: "Noun transfer",
+    initialState: ({ action }) => ({
+      nounId: action?.nounId || "",
+      receiverQuery: action?.target ?? "",
+    }),
+    useStateMiddleware: ({ state }) => {
+      const ensCache = useEnsCache();
+      const receiverQuery = state.receiverQuery?.trim();
+      const ensAddress = ensCache.resolve(receiverQuery);
+      const receiverAddress = isAddress(receiverQuery)
+        ? receiverQuery
+        : ensAddress ?? "";
+      const { address: treasuryAddress } =
+        getContractWithIdentifier("executor");
+      const totalSupply = useTotalSupply();
+      const maxNounId = totalSupply ? totalSupply - 2 : 0;
+      const treasuryNouns = useNounIdsByOwner({ owner: treasuryAddress });
+      return {
+        ...state,
+        receiverAddress,
+        treasuryAddress,
+        maxNounId,
+        treasuryNouns,
+      };
+    },
+    hasRequiredInputs: ({ state }) =>
+      state.nounId != null &&
+      state.receiverAddress != null &&
+      isAddress(state.receiverAddress) &&
+      state.treasuryNouns != null &&
+      state.treasuryNouns.includes(state.nounId),
+    buildAction: ({ state }) => ({
+      type: "treasury-noun-transfer",
+      target: state.receiverAddress,
+      nounId: state.nounId,
+    }),
+    Component: ({ state, setState }) => {
+      useCustomCacheEnsAddress(state.receiverQuery.trim(), {
+        enabled: state.receiverQuery.trim().split(".").slice(-1)[0].length > 0,
+      });
+      return (
+        <>
+          <div>
+            <Label htmlFor="nounId">Noun</Label>
+            <div
+              style={{
+                display: "grid",
+                gridTemplateColumns: "minmax(0,1fr) auto",
+                gap: "1rem",
+                alignItems: "center",
+              }}
+            >
+              <Input
+                type="number"
+                min={0}
+                max={state.maxNounId}
+                value={state.nounId}
+                onChange={(e) => {
+                  try {
+                    const n = BigInt(e.target.value);
+                    const truncatedN =
+                      n > state.maxNounId ? state.maxNounId : n < 0 ? 0 : n;
+                    setState({ nounId: truncatedN.toString() });
+                  } catch (e) {
+                    // Ignore
+                  }
+                }}
+                placeholder="0"
+              />
+              <NounAvatar id={state.nounId} size="3.5rem" />
+            </div>
+            <div
+              css={(t) =>
+                css({
+                  fontSize: t.text.sizes.small,
+                  color:
+                    state.nounId &&
+                    state.receiverAddress &&
+                    state.treasuryNouns &&
+                    !state.treasuryNouns.includes(state.nounId)
+                      ? t.colors.textHighlight
```

Comment:
always unsure about these lil ui things, switching colors and so on. if you have a better idea for this, I'm all ears